proposals/js-promise-integration/Overview.md

```diff
@@ -2,299 +2,225 @@
-   2. sets the state of `suspender` to **Active**[`caller`] (where `caller` is the current caller)
-   3. lets `result` be the result of calling `func(suspender,...args)` (or any trap or thrown exception)
-   4. asserts that `suspender`'s state is **Active**[`caller'`] for some `caller'` (should be guaranteed, though the caller might have changed)
-   5. changes `suspender`'s state to **Moribund**. This is also an opportunity to release any execution resources associated with the suspender. A **Moribund** suspender may not be used to suspend computations.
-   6. If `result` is not an exception or a trap, calls the `accept` function argument with the appropriate value.
-   6. If `result` is an exception, or if it is a trap, calls the `reject` function with the raised exception.
-6. Returns `promise` to `caller`
+We use the `Suspendable` constructor to signal to WebAssembly that a given import should cause a suspension of WebAssembly execution.
 
-Note that, if the inner function `func` suspends (by invoking a `Promise` returning import), then the `promise` will be returned to the `caller` before `func` returns. When `func` completes eventually, then `promise` will be resolved&mdash;and one of `accept` or `reject` will be invoked by the browser's microtask runner.
+#### `new WebAssembly.Suspending(`*`func`*`)`
 
-## Frequently Asked Questions
+The `WebAssembly.Suspending` constructor takes a JavaScript `Function` as an argument which is embedded in the `Suspending` object as the value of the `[[wrappedFunction]]` slot.
+
+>Note that *`func`* is expected to be a *JavaScript function*. This allows us to ignore certain so-called corner cases in the usage of JSPI: in particular there is no special handling of WebAssembly functions passed to `WebAssembly.Suspending`.
+
+1. If IsCallable(*func*) is `false`, throw a `TypeError` exception.
+1. Let *suspendingProto* be `WebAssembly.Suspendable.%prototype%`
+1. Let *susp* be the result of `OrdinaryObjectCreate`(*`suspendingProto`*)
+1. Assign the `[[wrappedFunction]]` internal slot of *`susp`* to *`func`*
+1. Return *susp*
+
+The most direct way that external functions can be accessed from a WebAssembly module is via imports.[^WebAssembly.Function] Therefore, we modify the *read-the-imports* algorithm to account for imports annotated as `Suspendable`.
+
+[^WebAssembly.Function]: The other way is by using a `WebAssembly.Function` constructor, as proposed in the js-types proposal. However, the semantics of `WebAssembly.Function` also revolve around modules and importing.
 
-1. **What is the purpose of the `Suspender` object?**
+We replace the section dealing with functions:
+
+3.4. If *externtype* is of the form `func` *functype* ...
+
+with:
+
+3.4. If *externtype* is of the form `func` *functype*
+
+  1. If `$IsCallable$`(*`v`*) is `true`
+      1. If *`v`* has a `FunctionAddress` internal slot, and therefore is an *Exported Function*,
```

Comment:
Here and below, `FunctionAddress` should be `[[FunctionAddress]]` (as it is in the published spec).

Reply:
I did wonder about that.